### ISSUE TITLE
[Feature]: Let users inject their own tracing layers

### ISSUE THREAD

**Issue by AlphaKeks** [shuttle-hq/shuttle#1003]
### Describe the feature

As of right now, shuttle uses a [Registry](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/registry/struct.Registry.html) to pick up logs from [tracing](https://docs.rs/tracing/latest/tracing/). This allows for attaching multiple [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html)s to the registry, which would then pick up the logs and can do whatever they want with them. Because shuttle sets up the registry, the user currently has no control over their own logs.

My idea was that we could add a parameter to the [#[shuttle_runtime::main]](https://docs.rs/shuttle-runtime/latest/shuttle_runtime/attr.main.html) macro that would allow the user to pass a function name that shuttle could then call to get one or more [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html)s to attach to the registry.

The main problem with this approach is that [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html) has a generic `S` which is the subscriber type it is attached to. I don't know if there is a way for the user to know that type ahead of time, so maybe we could expose a type alias from [shuttle_runtime](https://docs.rs/shuttle-runtime/latest/shuttle_runtime) that has the generic hard-coded according to whatever the default registry is?

### Suggestion or Example of how the feature would be used

The API for it could look something like this on the user side:

```rust
fn my_layer() -> Layer {
    let my_writer = todo!("...");

    tracing_subscriber::fmt::layer()
        .json()
        .with_writer(my_writer)
}

#[shuttle_runtime::main(tracing_layer = "my_layer")]
async fn init() -> shuttle_axum::ShuttleAxum {
    // ...
}
```

### Duplicate declaration

- [X] I have searched the issues and this feature has not been requested before.

**Comment by jonaro00** [comment]
How does this relate to #842 ? Compatible or replacement?

**Comment by AlphaKeks** [comment]
It would probably replace it. The goal of this is to give the user _full_ control, not just over the log level. I talked about this in the mentoring session this week and wasn't aware there was already an issue for it. I only searched for "tracing" and didn't find it - my bad. 

@unbekanntes-pferd what do you think about this? How much work is already done on your implementation?

**Comment by oddgrd** [comment]
So #842 addresses a concern we've had in the past about not being able to handle the volume of `trace` logs, because with that solution we can restrict the level a user can pass in. I still think the freedom your strategy adds is worth it, though, and I don't think a lot of projects will enable a `trace` level (and some my circumvent our log recorder anyway so it's not an issue). #842 would be easier to use for that simple use-case, but we could just document how to achieve the same thing with your solution. 

In conclusion I am leaning towards rather going with the solution proposed in this issue, but I will talk to Pieter and Iulian first and get back to you on that tomorrow.

Regarding your question about the layer generic, perhaps we can just [erase the type](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/index.html#runtime-configuration-with-layers) (see third paragraph) of the layer returned from the user?

**Comment by AlphaKeks** [comment]
I have already looked into this and wrote a small proof of concept just for myself. An idea that worked in my small contained environment was to just expose a concrete type the user could import, e.g.

```rust
// in `lib.rs` of `shuttle-runtime`

// or whatever type happens to be emitted by shuttle's defaults
pub type TracingLayer = Layered<Layer<Registry, Pretty, Format<Pretty>>, Registry>;
```

This would either require manual maintenance, which doesn't seem too bad considering this won't change often (hopefully), or we could also write some kind of macro that generates this type? :thinking: not sure about that yet.

**Comment by chesedo** [comment]
Won't a generic like this also work for the user code?

```rust
use tracing::info;
use tracing_subscriber::{Layer, Subscriber};

fn my_layer<S>() -> impl Layer<S> {
        tracing_subscriber::fmt::layer()
                .compact()
                .with_ansi(true)
}

#[shuttle_runtime::main(tracing_layer = my_layer)]
async fn init() -> ... {
        info!("Hey");
        // ...
}
```

**Comment by AlphaKeks** [comment]
The compiler wants functions to _always_ be valid, not just within the context they are actually used, so this leads to pretty ugly and verbose trait bounds. This is what the compiler wants me to do if I use a generic:

```rust
fn my_layer<S: tracing::Subscriber + for<'a> tracing_subscriber::registry::LookupSpan<'a>>(
) -> impl Layer<S> {
	tracing_subscriber::fmt::layer()
		.compact()
		.with_ansi(true)
}
```

Just to be clear: this works. But I think it's not very nice to look at. Also this might change if we change our subscriber which this will be attached to, not sure.

**Comment by that-ambuj** [comment]
I agree with @AlphaKeks here. Many people like their telemetry to be in a specific format like json and json lines for other observability platforms to use like Graphana and Promethius. Users should have full control over how much and what format of telemetry they collect from their applications.
